MOCSkyRegion : a region describing Multi-Order Coverage maps

bmatthieu3 opened the pull request:
Following #197 and #212 I moved a part of the code of [MOCPy](https://github.com/cds-astro/mocpy/) to astropy/regions .

This PR includes the definition of a MOCSkyRegion and MOCPixelRegion classes. Most of the code from MOCPy manipulating MOCs (classmethods, union, inter, contains...) is  moved to MOCSkyRegion.

I used a mpl PathPatch to plot the set of HEALPix cells describing a MOC.

The implementation is done following the principle that the user has to create a MOCSkyRegion first and then a pixel region and not the opposite. A MOCPixelRegion can be obtain from a sky region by calling its to_pixel method (the constructor of a pixel region accepts a sky region showing that a pixel region is deeply linked to a sky region). MOCPixelRegion.to_sky simply returns the sky region passed in the constructor of the pixel region.

There is a lot of code, I split it into several commits. The first one still has a lot of code but directly comes from MOCPy thus it has been tested by the CI tools of MOCPy (these tests have also been moved here and can be found in shapes/tests/test_interval_set.py and shapes/tests/test_moc.py files). The second to latest one are more linked to the integration in regions (i.e. doc, override of regions methods like to_pixel, contains, to_mask, bbox etc...).

Thank you a lot for reviewing it.

cdeil commented:
@bmatthieu3 - Thank you very much!

I'll try to find time to review this in the near future, apologies for not being responsive here.

@astrofrog or anyone - if you have time to do a review here, please do!

One first comment: I'd suggest to put the utility functions for `nuniq` handling that you have in utils here to astropy-healpix. It's e.g. also used in `hipspy`, where we don't use regions otherwise. If both `astropy.healpix` and `astropy.regions` are in the Astropy core, then it doesn't matter so much, but that's quite a way to go, and even then I think it fits better in `astropy.healpix`. So suggest to put it there first, and then import and use it from there in this PR.
See https://github.com/astropy/astropy-healpix/issues/91

cdeil commented on regions/shapes/moc/core.py:
```diff
@@ -0,0 +1,941 @@
+
+    Examples
+    --------
+    .. plot::
```
Personally I think a new high-level page `docs/moc.rst` that explains a bit what MOC is and gives examples like this one would be nicer than having long examples / plots in the API docs.
Note that it's easy to create links back and forth, for sure the class docstrings should contain a link to the high-level docs page.

cdeil commented on regions/shapes/moc/core.py:
```diff
@@ -0,0 +1,941 @@
+    @classmethod
+    def from_json(cls, json_moc, meta=None, visual=None):
+        """
+        Create a `~regions.MOCSkyRegion` from a dictionary of HEALPix cell arrays indexed by their order (i.e. in json format).
```
The first summary line of a docstring should always be short. In this case e.g. "Create from a JSON representation."

This line is super long and will not appear nicely in the method summary in the API docs.

Please build the HTML docs locally with `python setup.py build_docs` and browse through the API summary list and docstrings to make sure they look OK.

cdeil commented on regions/shapes/moc/core.py:
```diff
@@ -0,0 +1,941 @@
+        """
+        Returns the deepest order needed to describe self's interval set.
+        """
+        # TODO: cache value
```
What about this TODO? Please remove or address it.

(we should take this PR as an opportunity to improve and polish, once code has landed often this just doesn't happen any more; so please be a bit patient if I come with nitpicky comments in the coming days)

bmatthieu3 commented:
@cdeil - I did some changes in the last commits : 

* I added a moc.rst doc file containing two parts : an intro describing what the MOCs are and a part giving code example using regions to load a moc, plot it etc...
* I use the astropy-healpix ``level_to_nside``, ``uniq_to_level_ipix`` methods (repo version)
* Code refactoring (mainly renaming e.g. use of shorter variable names, remove the ``_arr`` suffix at the end of a numpy array, split the previous MOCSkyRegion.write method into two : a serialize and the write function. The first serializes the MOC into a HDU list tables/a python dict (JSON) and the last one stores the MOC into a file given a path.

One [test](https://travis-ci.org/astropy/regions/jobs/437216066) is still failing and I think it is maybe related to the release of matplotlib 3.0.0 ?

cdeil commented:
@bmatthieu3 - Is this ready to merge from your side?

Maybe have a look at https://coveralls.io/builds/19352514 and add a test exercising the few methods that don't have coverage?

I'm +1 to merge this in very soon and release it in v0.4 / advertise it to users to try out and give feedback.

@larrybradley @astrofrog @keflavich - Does one of you have time to do a review here?

bmatthieu3 commented:
@cdeil - I would like to take a little time to review my own code (it is from September and I think I did things in the past that I now do not do anymore) and import new features from what has been added in the current version of mocpy.
For example, there is:
- New methods for [plotting a moc and its perimeter(s)](https://github.com/cds-astro/mocpy/blob/master/notebooks/from_fits_and_intersection.ipynb). You can check the [binder hub](https://mybinder.org/v2/gh/cds-astro/mocpy/master) too. It can take some time to launch the image.
- A class WCS that allows the user to create more easily an astropy.wcs.WCS (for plotting purposes) from a center (skycoord), a fov (quantity), ...
- Some code refac, doc changes, optimizations.

I think this would not take too much time (between 1-2 weeks) to update the code and come back to you. Is this time frame ok for you @cdeil ?

cdeil commented:
@bmatthieu3 - Of course!

> * A class WCS that allows the user to create more easily an astropy.wcs.WCS (for plotting purposes) from a center (skycoord), a fov (quantity), ...

We have something similar in `hipspy`, but I don't think the API there is good. Maybe such a class could be shared somehow (I guess via Astropy core), or for now just copy & pasted to packages that want such a WCS maker helper, to avoid users having to learn different APIs for each package. If you think it's a good idea, and have a suggestion what to choose, please comment here: https://github.com/hipspy/hips/issues/82

bmatthieu3 commented:
Hi @cdeil - I enhanced the doc (new examples, plots added). 2 tests for travis-ci are still failing (the one using numpy 1.10 and [this one](https://travis-ci.org/astropy/regions/jobs/496401065)).
Coverage decreased because there is a lot of code that is involved in the plotting. Therefore, as matplotlib is not installed when testing, then the tests checking the plotting methods are not executed. Should I add matplotlib in the dependencies of travis (appveyor, circleci) ?

The appveyor job for win x64 py3.7 ([here](https://ci.appveyor.com/project/Astropy/regions/builds/22538489/job/4mllx1cartggya1l)) is also failing because [spherical-geometry](https://github.com/spacetelescope/spherical_geometry) does not build correctly. I need this package for creating a MOCSkyRegion from a polygon and also for plotting speedup optimizations. It is worth noting that the building of the spherical-geometry wheel for [win x64 py3.7](https://ci.appveyor.com/project/Astropy/regions/builds/22538489/job/4mllx1cartggya1l) fails whereas the one for [win x64 py2.7](https://ci.appveyor.com/project/Astropy/regions/builds/22538489/job/kaarmxix42vrko57) builds correctly.

bsipocz commented:
Failures, except the code style one, are unrelated. I'll try to make a generic maintenance PR  in the coming days (or of course others are welcome to do it,  too) so this can be then rebased on top.

bsipocz commented:
@bmatthieu3 - Could you try and rebase this on the latest master?

bmatthieu3 commented:
@bsipocz - I did the rebase and replaced some numpy calls that was not supported by numpy 1.10 (e.g. np.isin, np.flip). For the appveyor test, I opened an issue on spherical-geometry.

bsipocz commented:
@bmatthieu3 - Afaik they don't actively support windows.

bmatthieu3 commented:
@bsipocz - using VS 2017 made it work. I changed the appveyor.yml file in 
702d71b

astrofrog commented:
@bmatthieu3 @cdeil Just to check, is it correct that we decided at PyGamma19 not to go ahead with this but instead to refactor mocpy so that it depends on and inherits from the region classes here?

bmatthieu3 commented:
@cdeil @astrofrog Yes, the features implemented here are actually directly coming from mocpy. As for me, I would close this PR and keep enhancing mocpy instead.
